### PR TITLE
Add upe integration for the Checkout Block

### LINKED ISSUE #1667: SPIKE - UPE in Blocks Checkout
Currently the plugin [uses its own APIs to perform a block checkout](https://github.com/woocommerce/woocommerce-gateway-stripe/tree/develop/client/blocks). The goal of this spike is to provide details and a recommendation for building out the UPE block checkout using the plugin's APIs or using the [WC Blocks APIs](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/docs/extensibility/checkout-flow-and-events.md).

Some things to keep in mind:
* Determine the current interest/priority in moving existing gateways to the Blocks API
* Determine possibility of having the existing block checkout flow co-exist with a UPE Blocks checkout flow
* Determine scope/concerns of implementing the blocks checkout flow vs using the existing flow

### LINKED ISSUE #1787: UPE - Block checkout support
The goal of this issue is to bring support for UPE in the [block checkout](https://docs.woocommerce.com/document/using-the-new-block-based-checkout/).

The implementation will probably look similar to WCPay - 2429-gh-Automattic/woocommerce-payments.

Things to keep in mind:
- Make sure the existing express checkout block (aka payment request buttons) still works in the checkout flow.
- If #1730 is already implemented, make sure saving a payment method and using saved payment methods still works. Otherwise create a new issue for this feature.

### PR DESCRIPTION
**Pull request by reykjalin**
# Changes proposed in this Pull Request:

Fixes #1667 
Fixes #1787 

- When the new checkout experience (UPE) is enabled in settings, use the UPE element in the Checkout Block.

Parts of this PR cannot be tested and figured out until the UPE implementation in general supports Subscriptions and Pre-Orders.

I still expect parts of this PR to break once some of the in-progress PRs are merged, but this is ready to review with the current status in `develop`.

# Testing instructions

Go through the following critical flows with the UPE experience enabled and using the Checkout block on the checkout page.

## With UPE disabled

We've run into some issues when UPE has been disabled, so please run through the following critical flows with UPE disabled, using both the Blocks and Shortcode versions of the checkout flow.

* [Checkout with normal credit card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#checkout-failures-with-various-cards#checkout-with-normal-credit-card)
* [Save card at checkout](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#checkout-failures-with-various-cards#save-card-at-checkout)
* [Checkout with saved card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#checkout-failures-with-various-cards#checkout-with-saved-card)
* [Checkout with SCA card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#checkout-failures-with-various-cards#checkout-with-sca-card)
* [Checkout failures (with various cards)](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#checkout-failures-with-various-cards#checkout-failures-with-various-cards)
* [Checkout with Payment Request](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#checkout-failures-with-various-cards#checkout-with-payment-request)
* [Blocks: Checkout with normal credit card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-checkout-with-normal-credit-card)
* [Blocks: Checkout with SCA card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-checkout-with-sca-card)
* [Blocks: Checkout failures (with various cards)](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-checkout-failures-with-various-cards)
* [Blocks: Save card at checkout](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-save-card-at-checkout)
* [Blocks: Checkout with saved card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-checkout-with-saved-card)
* [Blocks: Checkout with Payment Request](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-checkout-with-payment-request-button)

## Gutenberg

- [Blocks: Create a Checkout page with the Checkout Block](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-create-a-checkout-page-with-the-checkout-block)

## Adding/deleting payment methods

This PR makes some changes to the structure of the JS and shifts many functions to new files. Please make sure the non-checkout based functionality that uses the UPE element also works:

* [Add new payment method](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#add-new-payment-method#add-new-payment-method)
* [Delete payment method](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#add-new-payment-method#delete-payment-method)

## Card payments

Please make sure to test both the Shortcode and Blocks-based checkout flows.

### Shortcode flow

* [Checkout with normal credit card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#checkout-failures-with-various-cards#checkout-with-normal-credit-card)
* [Save card at checkout](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#checkout-failures-with-various-cards#save-card-at-checkout)
* [Checkout with saved card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#checkout-failures-with-various-cards#checkout-with-saved-card)
* [Checkout with SCA card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#checkout-failures-with-various-cards#checkout-with-sca-card)
* [Checkout failures (with various cards)](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#checkout-failures-with-various-cards#checkout-failures-with-various-cards)
* [Checkout with Payment Request](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#checkout-failures-with-various-cards#checkout-with-payment-request)

### Blocks flow

* [Blocks: Checkout with normal credit card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-checkout-with-normal-credit-card)
* [Blocks: Checkout with SCA card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-checkout-with-sca-card)
* [Blocks: Checkout failures (with various cards)](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-checkout-failures-with-various-cards)
* [Blocks: Save card at checkout](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-save-card-at-checkout)
* [Blocks: Checkout with saved card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-checkout-with-saved-card)
* [Blocks: Checkout with Payment Request](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-checkout-with-payment-request-button)

## Payment Request Buttons

Depends on:

- [x] #1908 

* [Blocks: Checkout with Payment Request](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-checkout-with-payment-request-button)

## Subscriptions

Depends on:

- [ ] #1836 

* [Blocks: Purchase subscription product](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-purchase-subscription-product)
* [Blocks: Purchase free trial subscription](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-purchase-free-trial-subscription)
* [Blocks: Purchase multiple subscriptions](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-purchase-multiple-subscriptions)
* [Renew subscription manually](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#renew-subscription-manually)
* [Renew subscription automatically](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#renew-subscription-automatically)
* [Renew subscription](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#renew-subscription)

-------------------
- [x] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [x] Did you make changes, or create a **new .js file**? If **Gruntfile.js** exists in the repo, make sure to run `grunt`.

### CODE REVIEW

File: client/blocks/upe/confirm-card-payment.js
Comment (reykjalin) [inline client/blocks/upe/confirm-card-payment.js:22]: This works even though `paymentMethod` is always `undefined` here, probably since we only need the payment method ID if we're actually saving the payment method? So maybe we can just replace the `shouldSavePayment ? paymentMethod : null` with `null` or `undefined`?

I don't fully understand what the API expects here, so I'm at a bit of a loss here; this is all copied from WCPay.

I'm honestly not really sure what this parameter is for

File: includes/class-wc-stripe-blocks-support.php
Comment (reykjalin) [inline includes/class-wc-stripe-blocks-support.php]: This seems extremely backwards, but without this the supports array is empty when using UPE-based checkout. That would be a problem when we add support for Subscriptions and Pre-Orders when using UPE.

Ideally this should be removed, and I'm hoping the UPE gateway will become part of the `$gateways` array once #1856 is merged.

File: .eslintrc.js
Comment (reykjalin) [inline .eslintrc.js]: This is to fix an issue with ESLint. Should probably be part of #1854

File: package.json
Comment (reykjalin) [inline package.json]: This is to fix an issue with ESLint. Should probably be part of #1854

File: client/blocks/upe/fields.js
Comment (reykjalin) [inline client/blocks/upe/fields.js]: I'm pretty sure these shouldn't all be set to `'never'`, but instead use the `enabledBillingFields` setting we embed from the PHP side. Not sure how that plays with the Blocks integration though; that's something we need to look into.

File: includes/class-wc-stripe-blocks-support.php
Comment (reykjalin) [inline includes/class-wc-stripe-blocks-support.php]: I think this is a better approach to loading the relevant JS because it allows us to load smaller JS bundles.

The alternative would be to have the JS decide whether the UPE or "Classic" experience should be loaded, but that would mean a bigger JS bundle since the JS would contain both the "Classic" and UPE experiences; roughly half of the bundle would go unused in this case.

The downside of this approach is that the built plugin assets will be a bit bigger so merchants will need to download a slightly bigger plugin archive when installing/updating the plugin. I think the trade-off here is worth it though; a lighter (and thus faster) checkout experience is always worth it.

That said, I'm still open to suggestions on this 🙂

File: includes/class-wc-stripe-blocks-support.php
Comment (reykjalin) [inline includes/class-wc-stripe-blocks-support.php]: I'm hoping we can remove this once #1855 has been merged since the classic UPE JS (`upe_classic.js`) has some sort of conflict with the Blocks integration (`upe_blocks.js`).

**Note:** this (very simple) workaround breaks the shortcode checkout flow, so we should not merge this PR until this has been resolved.

File: client/blocks/upe/fields.js
Comment (diegocurbelo) [inline client/blocks/upe/fields.js]: We should use `__()`:
```
__( 'There was an error loading the payment gateway', 'woocommerce-gateway-stripe' )
```

File: client/blocks/upe/fields.js
Comment (diegocurbelo) [inline client/blocks/upe/fields.js]: We should use `__()`:
```
message: __( 'Your payment information is incomplete.', 'woocommerce-gateway-stripe' ),
```

File: client/blocks/upe/fields.js
Comment (diegocurbelo) [inline client/blocks/upe/fields.js:94]: We should use __():
```
__( 'This payment method can not be saved for future use.', 'woocommerce-gateway-stripe' ), 
```

File: includes/class-wc-stripe-blocks-support.php
Comment (diegocurbelo) [inline includes/class-wc-stripe-blocks-support.php]: I think that `classic` is a bit confusing here, generally, we refer to the shortcode checkout version as classic, like in `upe_classic.js`, and here we are registering the normal (non-UPE) version of the blocks integration.

But I don't have a better idea than `register_non_upe_payment_method_script_handles` or `register_legacy_payment_method_script_handles `😞 .

Maybe the problem es using `classic` for the shortcode and we need to change that 🤷🏼

File: includes/class-wc-stripe-blocks-support.php
Comment (diegocurbelo) [reply includes/class-wc-stripe-blocks-support.php]: Can't reproduce this behavior, debugged this for a while, and the UPE Gateway is also called `stripe` so it's always set and returns in the first if.

We can however change it to not access the property `payment_gateways ` directly:
```
$gateways = WC()->payment_gateways()->get_available_payment_gateways();
```

**Review by diegocurbelo — CHANGES_REQUESTED** [review]
Code looks good, and test very well 👏🏼 👏🏼 . Just added a couple of comments about missing translations.

I'm going to double check the `wp_dequeue_script( 'wc-stripe-upe-classic' );` that I can't reproduce.

File: client/blocks/upe/fields.js
Comment (reykjalin) [reply client/blocks/upe/fields.js:94]: Fixed in c8c04c83ca6989971a45ffea9d9079eee7fbc177

File: client/blocks/upe/fields.js
Comment (reykjalin) [reply client/blocks/upe/fields.js]: Fixed in c8c04c83ca6989971a45ffea9d9079eee7fbc177

File: client/blocks/upe/fields.js
Comment (reykjalin) [reply client/blocks/upe/fields.js]: Fixed in c8c04c83ca6989971a45ffea9d9079eee7fbc177

File: includes/class-wc-stripe-blocks-support.php
Comment (reykjalin) [reply includes/class-wc-stripe-blocks-support.php]: I like `legacy`; seems pretty accurate since the aim is to _eventually_ switch entirely to UPE 🤔 

Changed it to `legacy` in a67046de

File: includes/class-wc-stripe-blocks-support.php
Comment (reykjalin) [reply includes/class-wc-stripe-blocks-support.php]: This has been removed since #1855 has been merged.

The issue itself also turned out to be a bit more nuanced, as can be seen with https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1825/commits/a9774f9f6745f8effaa9d3fb0df6d133248999fa

File: includes/class-wc-stripe-blocks-support.php
Comment (reykjalin) [reply includes/class-wc-stripe-blocks-support.php]: Yep, I can no longer reproduce this after #1855 was merged, so I've removed this code 🙂

File: client/stripe-utils/utils.js
Comment (reykjalin) [inline client/stripe-utils/utils.js]: This function is now exclusively used for the UPE checkout, since it was causing a conflict with the Blocks integration. See the commit message in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1825/commits/a9774f9f6745f8effaa9d3fb0df6d133248999fa for more details.

File: client/blocks/utils.js
Comment (reykjalin) [inline client/blocks/utils.js]: This function is now exclusively used for the Blocks integration, since the `getStripeServerData()` function in `stripe-utils` was causing a conflict with the UPE integration. See the commit message in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1825/commits/a9774f9f6745f8effaa9d3fb0df6d133248999fa for more details.

I decided to go with the same name since it serves the same purpose as the `getStripeServerData()` function in `stripe-utils`, but that may be confusing. Would it make more sense to call this function `getBlocksConfiguration()` (or similar) so it doesn't get confused with the other `getStripeServerData()` function?

Comment (diegocurbelo) [reply client/blocks/utils.js]: I like the idea of having different names... but in that case should we also rename the classic `getStripeServerData()` (the one in `client/stripe-utils/utils.js`) `getClassicConfiguration()`?

Comment (reykjalin) [reply client/blocks/utils.js]: I think so? But I'm worried about messing up other people's PRs that way, so I think we should open an issue for that 🤔 

For now I can change the Blocks integration function to `getBlocksConfiguration()` to reduce the ambiguity 👍

Comment (reykjalin) [reply client/blocks/utils.js]: Fixed in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1825/commits/a9e2c1955077f82046623692bdea54039c9c94a4

**Comment by reykjalin** [comment]
@diegocurbelo the PR is ready for another review, but unfortunately parts of it can't be tested because of dependencies on other PRs and issues.

I think it's probably worth it to review the code and make sure the tests that work pass so there's less to do next time, but it may also be better to simply wait. I'll ping you again no matter what you decide to do once the PR can be reviewed more thoroughly 🙂

File: client/api/blocks.js
Comment (reykjalin) [inline client/api/blocks.js:1]: This file was added to separate the UPE/combined API calls entirely from the Blocks API calls.

File: client/blocks/utils.js
Comment (reykjalin) [inline client/blocks/utils.js:1]: These are utility functions exclusive to the Blocks integration.

File: client/stripe-utils/utils.js
Comment (diegocurbelo) [inline client/stripe-utils/utils.js:22]: Reverting this solves the `wc_stripe_upe_params is not defined` error
Screen Shot 2021-09-22 at 17 12 52
:

**Review by diegocurbelo — APPROVED** [review]

**Pull request merged**
